# Superhero

- Clint Atterberry
- 8/29/22

## Project Information

The Task
Your task is two-fold:

I. Clean the files and combine them into one final DataFrame.

    - This dataframe should have the following columns:
        - Hero (Just the name of the Hero)
        - Publisher
        - Gender
        - Eye color
        - Race
        - Hair color
        - Height (numeric)
        - Skin color
        - Alignment
        - Weight (numeric)
        - Plus, one-hot-encoded columns for every power that - appears in the dataset. E.g.:
            - Agility
            - Flight
            - Superspeed
            - etc.
Hint: There is a space in "100 kg" or "52.5 cm"



II. Use your combined DataFrame to answer the following questions.

1. Compare the average weight of super powers who have Super Speed to those who do not.
2. What is the average height of heroes for each publisher?

# Setup

## Import

In [1]:
# Standard Imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Importing the OS and JSON Modules
import os,json

## Load Data

In [2]:
super_info = 'https://docs.google.com/spreadsheets/d/e/2PACX-1vQqIOAkjE3NPqH3fN5qdMIyEMIAbBXMyHHYgjEdDfgttcU0dbl8dCAHze2iestVkaSbRKSXS3YpkbFT/pub?output=csv'
super_powers = 'https://docs.google.com/spreadsheets/d/e/2PACX-1vSZeRKvu4AbsTWiYbk3mHKTOjAHlIlyCKOcpliC39eCELYeLOgKMmDN2AdEohkxEWNCvvrPiJ26Acop/pub?output=csv'
super_info_df = pd.read_csv(super_info)
super_powers_df = pd.read_csv(super_powers)

## Explore Data

In [3]:
# looking at the first 3 rows of data
super_info_df.head(3)

Hero|Publisher Gender           Race Alignment Hair color  \
0          A-Bomb|Marvel Comics   Male          Human      good    No Hair   
1  Abe Sapien|Dark Horse Comics   Male  Icthyo Sapien      good    No Hair   
2            Abin Sur|DC Comics   Male        Ungaran      good    No Hair   

  Eye color Skin color                                  Measurements  
0    yellow    Unknown  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
1      blue       blue   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  
2      blue        red   {'Height': '185.0 cm', 'Weight': '90.0 kg'}

In [4]:
# looking at the first 3 rows of data
super_powers_df.head(3)

hero_names                                             Powers
0     3-D Man         Agility,Super Strength,Stamina,Super Speed
1      A-Bomb  Accelerated Healing,Durability,Longevity,Super...
2  Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...

In [5]:
# explore super_info
super_info_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Hero|Publisher  463 non-null    object
 1   Gender          463 non-null    object
 2   Race            463 non-null    object
 3   Alignment       463 non-null    object
 4   Hair color      463 non-null    object
 5   Eye color       463 non-null    object
 6   Skin color      463 non-null    object
 7   Measurements    463 non-null    object
dtypes: object(8)
memory usage: 29.1+ KB


- 8 columns
- 463 rows
- Appears to have no missing data in the columns
- All columns are objects
- Measurements column will need to be numeric and separated into 2 columns, height and weight
- Hero and Publisher will need to be separated into 2 columns

In [6]:
# explore super_powers
super_powers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 667 entries, 0 to 666
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   hero_names  667 non-null    object
 1   Powers      667 non-null    object
dtypes: object(2)
memory usage: 10.5+ KB


- 2 columns
- 667 rows
- Appears to have no missing data in the columns
- Both columns are objects
- Powers will need to be one-hot-encoded to different columns

# Prepare Data (super_info_df)

## Separate Hero and Publisher

In [7]:
# Exploring existing format with a few examples
super_info_df['Hero|Publisher'].head(2)

0            A-Bomb|Marvel Comics
1    Abe Sapien|Dark Horse Comics
Name: Hero|Publisher, dtype: object

The '|' separates the two columns.

In [8]:
## adding expand=True to create the 2 columns
super_info_df[['Hero','Publisher']] = super_info_df['Hero|Publisher'].str.split('|',expand=True)
super_info_df.head(2)

Hero|Publisher Gender           Race Alignment Hair color  \
0          A-Bomb|Marvel Comics   Male          Human      good    No Hair   
1  Abe Sapien|Dark Horse Comics   Male  Icthyo Sapien      good    No Hair   

  Eye color Skin color                                  Measurements  \
0    yellow    Unknown  {'Height': '203.0 cm', 'Weight': '441.0 kg'}   
1      blue       blue   {'Height': '191.0 cm', 'Weight': '65.0 kg'}   

         Hero          Publisher  
0      A-Bomb      Marvel Comics  
1  Abe Sapien  Dark Horse Comics

In [9]:
super_info_df.drop(columns='Hero|Publisher', inplace=True)
super_info_df.head(2)

Gender           Race Alignment Hair color Eye color Skin color  \
0   Male          Human      good    No Hair    yellow    Unknown   
1   Male  Icthyo Sapien      good    No Hair      blue       blue   

                                   Measurements        Hero          Publisher  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}      A-Bomb      Marvel Comics  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  Abe Sapien  Dark Horse Comics

The column was successfully separated into two different columns.

## Separate Height and Weight

In [10]:
# Exploring existing format with a few examples
super_info_df['Measurements'].head(2)

0    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
1     {'Height': '191.0 cm', 'Weight': '65.0 kg'}
Name: Measurements, dtype: object

The keys need to be columns, and the values will be items in those respective columns.

In [11]:
## examining a single value from the 'measurements' col
height_and_weight = super_info_df.loc[0,'Measurements']
print(type(height_and_weight))
height_and_weight

<class 'str'>


"{'Height': '203.0 cm', 'Weight': '441.0 kg'}"

- Notice that there are quotation marks around the entire dictionary! We actually have a STRING that LOOKS like a dictionary.
- The entire coordinates column is dictionaries-turned-strings that must be converted back to dictionaries.

In [12]:
# The data will not be able to be cleaned because of the double quotes " "
# To fix this, we can replace our single quotes with double quotes before we use json.loads.
# For our single test string, we can use the .replace method:

height_and_weight = height_and_weight.replace("'",'"')
height_and_weight

'{"Height": "203.0 cm", "Weight": "441.0 kg"}'

In [13]:
## now we can use json.loads
fixed_height_and_weight = json.loads(height_and_weight)
print(type(fixed_height_and_weight))
fixed_height_and_weight

<class 'dict'>


{'Height': '203.0 cm', 'Weight': '441.0 kg'}

Now that we know this will work, we can apply it to the entire column.

We can apply this transformation to the entire column by combining:
- Using .str.replace to replace single quotes.
- Using .apply with json.loads to convert all rows simultaneously

In [14]:
## use .str.replace to replace all single quotes
super_info_df['Measurements'] = super_info_df['Measurements'].str.replace("'",'"')
## Apply the json.loads to the full column
super_info_df['Measurements'] = super_info_df['Measurements'].apply(json.loads)
super_info_df['Measurements'].head()

0    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
1     {'Height': '191.0 cm', 'Weight': '65.0 kg'}
2     {'Height': '185.0 cm', 'Weight': '90.0 kg'}
3    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
4    {'Height': '193.0 cm', 'Weight': '122.0 kg'}
Name: Measurements, dtype: object

In [15]:
## check a single value after transformation
test_measurement = super_info_df.loc[0, 'Measurements']
print(type(test_measurement))
test_measurement

<class 'dict'>


{'Height': '203.0 cm', 'Weight': '441.0 kg'}

- We now want to convert the single "Measurements" column into 2 separate columns "Height" and "Weight".
    - To do so, we will use .apply with the pd.Series function.

- If we use the pandas pd.Series function on our column of dictionaries, it will automatically convert each key to a new column!

In [16]:
height_weight = super_info_df['Measurements'].apply(pd.Series)
height_weight

Height    Weight
0    203.0 cm  441.0 kg
1    191.0 cm   65.0 kg
2    185.0 cm   90.0 kg
3    203.0 cm  441.0 kg
4    193.0 cm  122.0 kg
..        ...       ...
458  183.0 cm   83.0 kg
459  165.0 cm   52.0 kg
460   66.0 cm   17.0 kg
461  170.0 cm   57.0 kg
462  185.0 cm   81.0 kg

[463 rows x 2 columns]

Now, we concatenate these 2 new columns to our original dataframe AND drop the original column.
- We will use pd.concat with axis=1 to concatenate the two dataframes side by side.
    - pd.concat([df1,df2], axis=1)

In [17]:
# concat height_weight with original dataframe
super_info_df = pd.concat([super_info_df, height_weight], axis = 1)
super_info_df.head(2)

Gender           Race Alignment Hair color Eye color Skin color  \
0   Male          Human      good    No Hair    yellow    Unknown   
1   Male  Icthyo Sapien      good    No Hair      blue       blue   

                                   Measurements        Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}      A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  Abe Sapien   

           Publisher    Height    Weight  
0      Marvel Comics  203.0 cm  441.0 kg  
1  Dark Horse Comics  191.0 cm   65.0 kg

Success! Now to drop the 'Measurements' column.

In [18]:
# dropping the 'Measurements' column
super_info_df.drop(columns='Measurements', inplace=True)
super_info_df.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

            Hero          Publisher    Height    Weight  
0         A-Bomb      Marvel Comics  203.0 cm  441.0 kg  
1     Abe Sapien  Dark Horse Comics  191.0 cm   65.0 kg  
2       Abin Sur          DC Comics  185.0 cm   90.0 kg  
3    Abomination      Marvel Comics  203.0 cm  441.0 kg  
4  Absorbing Man      Marvel Comics  193.0 cm  122.0 kg

## Change dtype of 'Height' and 'Weight'

In [19]:
# checking dtypes
super_info_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Gender      463 non-null    object
 1   Race        463 non-null    object
 2   Alignment   463 non-null    object
 3   Hair color  463 non-null    object
 4   Eye color   463 non-null    object
 5   Skin color  463 non-null    object
 6   Hero        463 non-null    object
 7   Publisher   463 non-null    object
 8   Height      463 non-null    object
 9   Weight      463 non-null    object
dtypes: object(10)
memory usage: 36.3+ KB


Height and Weight are both objects still, need to change to numeric.

In [20]:
# checking for the values, if they are int or float
print(super_info_df['Height'].head(10))
print(super_info_df['Weight'].head(10))

0    203.0 cm
1    191.0 cm
2    185.0 cm
3    203.0 cm
4    193.0 cm
5    185.0 cm
6    178.0 cm
7    191.0 cm
8    188.0 cm
9    193.0 cm
Name: Height, dtype: object
0    441.0 kg
1     65.0 kg
2     90.0 kg
3    441.0 kg
4    122.0 kg
5     88.0 kg
6     81.0 kg
7    104.0 kg
8    108.0 kg
9     90.0 kg
Name: Weight, dtype: object


Converting the values to int because the decimals are all '0'.

In [21]:
# pd.to_numeric(super_info_df['Height'], downcast="integer")
# pd.to_numeric(super_info_df['Weight'], downcast="integer")

# super_info_df.info()

Error! Need to get rid of the " cm" and " kg".

In [22]:
super_info_df.replace([' cm', ' kg'], '', inplace=True, regex=True)
super_info_df.head(2)

Gender           Race Alignment Hair color Eye color Skin color        Hero  \
0   Male          Human      good    No Hair    yellow    Unknown      A-Bomb   
1   Male  Icthyo Sapien      good    No Hair      blue       blue  Abe Sapien   

           Publisher Height Weight  
0      Marvel Comics  203.0  441.0  
1  Dark Horse Comics  191.0   65.0

Success! Checking the dtype again.

In [23]:
super_info_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Gender      463 non-null    object
 1   Race        463 non-null    object
 2   Alignment   463 non-null    object
 3   Hair color  463 non-null    object
 4   Eye color   463 non-null    object
 5   Skin color  463 non-null    object
 6   Hero        463 non-null    object
 7   Publisher   463 non-null    object
 8   Height      463 non-null    object
 9   Weight      463 non-null    object
dtypes: object(10)
memory usage: 36.3+ KB


Still 'obj', will now change to 'float'.

In [24]:
super_info_df['Height'] = pd.to_numeric(super_info_df['Height'])
super_info_df['Weight'] = pd.to_numeric(super_info_df['Weight'])
super_info_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Gender      463 non-null    object 
 1   Race        463 non-null    object 
 2   Alignment   463 non-null    object 
 3   Hair color  463 non-null    object 
 4   Eye color   463 non-null    object 
 5   Skin color  463 non-null    object 
 6   Hero        463 non-null    object 
 7   Publisher   463 non-null    object 
 8   Height      463 non-null    float64
 9   Weight      463 non-null    float64
dtypes: float64(2), object(8)
memory usage: 36.3+ KB


Success!

# Prepare Data (super_powers_df)

## Separate the 'Powers' into their own columns

In [25]:
# review super_powers_df
super_powers_df.head()

hero_names                                             Powers
0      3-D Man         Agility,Super Strength,Stamina,Super Speed
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...
3     Abin Sur                                 Lantern Power Ring
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...

In [26]:
## showing the lists are really strings
super_powers_df.loc[2,'Powers']

'Agility,Accelerated Healing,Cold Resistance,Durability,Underwater breathing,Marksmanship,Weapons Master,Longevity,Intelligence,Super Strength,Telepathy,Stamina,Immortality,Reflexes,Enhanced Sight,Sub-Mariner'

In [27]:
# Create a new column where the single quotes are replaced by double quotes
super_powers_df['Powers_Split'] = super_powers_df['Powers'].str.split(',', expand=False)

In [28]:
## showing the lists are really strings
super_powers_df.loc[2,'Powers_Split']

['Agility',
 'Accelerated Healing',
 'Cold Resistance',
 'Durability',
 'Underwater breathing',
 'Marksmanship',
 'Weapons Master',
 'Longevity',
 'Intelligence',
 'Super Strength',
 'Telepathy',
 'Stamina',
 'Immortality',
 'Reflexes',
 'Enhanced Sight',
 'Sub-Mariner']

In [29]:
# exploring the .value_counts in the column 'Powers_Split'
super_powers_df['Powers_Split'].value_counts()

[Intelligence]                                                                                                                                                                                                                                                                          8
[Durability, Super Strength]                                                                                                                                                                                                                                                            5
[Agility, Stealth, Marksmanship, Weapons Master, Stamina]                                                                                                                                                                                                                               4
[Marksmanship]                                                                                                                                            

In [30]:
## exploding the column of lists
exploded = super_powers_df.explode('Powers_Split')
exploded[['hero_names','Powers','Powers_Split']].head(5)

hero_names                                             Powers  \
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
1     A-Bomb  Accelerated Healing,Durability,Longevity,Super...   

          Powers_Split  
0              Agility  
0       Super Strength  
0              Stamina  
0          Super Speed  
1  Accelerated Healing

In [31]:
## saving the unique values from the exploded column
cols_to_make = exploded['Powers_Split'].dropna().unique()
cols_to_make

array(['Agility', 'Super Strength', 'Stamina', 'Super Speed',
       'Accelerated Healing', 'Durability', 'Longevity', 'Camouflage',
       'Self-Sustenance', 'Cold Resistance', 'Underwater breathing',
       'Marksmanship', 'Weapons Master', 'Intelligence', 'Telepathy',
       'Immortality', 'Reflexes', 'Enhanced Sight', 'Sub-Mariner',
       'Lantern Power Ring', 'Invulnerability', 'Animation',
       'Super Breath', 'Dimensional Awareness', 'Flight', 'Size Changing',
       'Teleportation', 'Magic', 'Dimensional Travel',
       'Molecular Manipulation', 'Energy Manipulation', 'Power Cosmic',
       'Energy Absorption', 'Elemental Transmogrification',
       'Fire Resistance', 'Natural Armor', 'Heat Resistance',
       'Matter Absorption', 'Regeneration', 'Stealth', 'Power Suit',
       'Energy Blasts', 'Energy Beams', 'Heat Generation', 'Danger Sense',
       'Phasing', 'Force Fields', 'Hypnokinesis', 'Invisibility',
       'Enhanced Senses', 'Jump', 'Shapeshifting', 'Elasticity',
 

In [32]:
# using a For Loop and .str.contains to create the new columns
for col in cols_to_make:
    super_powers_df[col] = super_powers_df['Powers'].str.contains(col)
super_powers_df.head()

C:\Users\oneda\AppData\Local\Temp\ipykernel_2400\2407499648.py:3: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  super_powers_df[col] = super_powers_df['Powers'].str.contains(col)
C:\Users\oneda\AppData\Local\Temp\ipykernel_2400\2407499648.py:3: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  super_powers_df[col] = super_powers_df['Powers'].str.contains(col)
C:\Users\oneda\AppData\Local\Temp\ipykernel_2400\2407499648.py:3: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many

C:\Users\oneda\AppData\Local\Temp\ipykernel_2400\2407499648.py:3: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  super_powers_df[col] = super_powers_df['Powers'].str.contains(col)
C:\Users\oneda\AppData\Local\Temp\ipykernel_2400\2407499648.py:3: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  super_powers_df[col] = super_powers_df['Powers'].str.contains(col)
C:\Users\oneda\AppData\Local\Temp\ipykernel_2400\2407499648.py:3: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many

C:\Users\oneda\AppData\Local\Temp\ipykernel_2400\2407499648.py:3: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  super_powers_df[col] = super_powers_df['Powers'].str.contains(col)
C:\Users\oneda\AppData\Local\Temp\ipykernel_2400\2407499648.py:3: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  super_powers_df[col] = super_powers_df['Powers'].str.contains(col)
C:\Users\oneda\AppData\Local\Temp\ipykernel_2400\2407499648.py:3: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many

hero_names                                             Powers  \
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
3     Abin Sur                                 Lantern Power Ring   
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...   

                                        Powers_Split  Agility  Super Strength  \
0    [Agility, Super Strength, Stamina, Super Speed]     True            True   
1  [Accelerated Healing, Durability, Longevity, S...    False            True   
2  [Agility, Accelerated Healing, Cold Resistance...     True            True   
3                               [Lantern Power Ring]    False           False   
4  [Accelerated Healing, Intelligence, Super Stre...    False            True   

   Stamina  Super Speed  Accelerated Healing  Durability  Longevity  ...  \
0     True         True                False       False      False  ...   
1     True        False                 True        True       True  ...   
2     True        False                 True        True       True  ...   
3    False        False                False       False      False  ...   
4     True         True                 True       False      False  ...   

   Weather Control  Omnipresent  Omniscient  Hair Manipulation  Nova Force  \
0            False        False       False              False       False   
1            False        False       False              False       False   
2            False        False       False              False       False   
3            False        False       False              False       False   
4            False        False       False              False       False   

   Odin Force  Phoenix Force  Intuitive aptitude  Melting  Changing Armor  
0       False          False               False    False           False  
1       False          False               False    False           False  
2       False          False               False    False           False  
3       False          False               False    False           False  
4       False          False               False    False           False  

[5 rows x 170 columns]

Success!

In [33]:
# explore super_powers_df to view columns
super_powers_df.head()

hero_names                                             Powers  \
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
3     Abin Sur                                 Lantern Power Ring   
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...   

                                        Powers_Split  Agility  Super Strength  \
0    [Agility, Super Strength, Stamina, Super Speed]     True            True   
1  [Accelerated Healing, Durability, Longevity, S...    False            True   
2  [Agility, Accelerated Healing, Cold Resistance...     True            True   
3                               [Lantern Power Ring]    False           False   
4  [Accelerated Healing, Intelligence, Super Stre...    False            True   

   Stamina  Super Speed  Accelerated Healing  Durability  Longevity  ...  \
0     True         True                False       False      False  ...   
1     True        False                 True        True       True  ...   
2     True        False                 True        True       True  ...   
3    False        False                False       False      False  ...   
4     True         True                 True       False      False  ...   

   Weather Control  Omnipresent  Omniscient  Hair Manipulation  Nova Force  \
0            False        False       False              False       False   
1            False        False       False              False       False   
2            False        False       False              False       False   
3            False        False       False              False       False   
4            False        False       False              False       False   

   Odin Force  Phoenix Force  Intuitive aptitude  Melting  Changing Armor  
0       False          False               False    False           False  
1       False          False               False    False           False  
2       False          False               False    False           False  
3       False          False               False    False           False  
4       False          False               False    False           False  

[5 rows x 170 columns]

In [34]:
# dropping 'Powers' and 'Powers_Split', not needed
super_powers_df.drop(columns = (['Powers', 'Powers_Split']), inplace=True)
super_powers_df.head(2)

hero_names  Agility  Super Strength  Stamina  Super Speed  \
0    3-D Man     True            True     True         True   
1     A-Bomb    False            True     True        False   

   Accelerated Healing  Durability  Longevity  Camouflage  Self-Sustenance  \
0                False       False      False       False            False   
1                 True        True       True        True             True   

   ...  Weather Control  Omnipresent  Omniscient  Hair Manipulation  \
0  ...            False        False       False              False   
1  ...            False        False       False              False   

   Nova Force  Odin Force  Phoenix Force  Intuitive aptitude  Melting  \
0       False       False          False               False    False   
1       False       False          False               False    False   

   Changing Armor  
0           False  
1           False  

[2 rows x 168 columns]

# Merge Dataframes

## Joining Dataframes

In [35]:
# renaming the column 'hero_names' to 'Hero' for consistency
super_powers_df.rename(columns={'hero_names':'Hero'}, inplace=True)

## Merging with the same column name
super_hero_df = pd.merge(super_info_df,super_powers_df, on='Hero')
super_hero_df.head(2)

Gender           Race Alignment Hair color Eye color Skin color        Hero  \
0   Male          Human      good    No Hair    yellow    Unknown      A-Bomb   
1   Male  Icthyo Sapien      good    No Hair      blue       blue  Abe Sapien   

           Publisher  Height  Weight  ...  Weather Control  Omnipresent  \
0      Marvel Comics   203.0   441.0  ...            False        False   
1  Dark Horse Comics   191.0    65.0  ...            False        False   

   Omniscient  Hair Manipulation  Nova Force  Odin Force  Phoenix Force  \
0       False              False       False       False          False   
1       False              False       False       False          False   

   Intuitive aptitude  Melting  Changing Armor  
0               False    False           False  
1               False    False           False  

[2 rows x 177 columns]

In [36]:
super_hero_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 463 entries, 0 to 462
Columns: 177 entries, Gender to Changing Armor
dtypes: bool(167), float64(2), object(8)
memory usage: 115.3+ KB


# Solutions to Questions

## Question 1

In [37]:
# filtering the super_hero_df for Super Speed == True and false
super_speed_true = super_hero_df[super_hero_df['Super Speed']== True]
super_speed_false = super_hero_df[super_hero_df['Super Speed']== False]
print(f"The mean weight for those with Super Speed is {super_speed_true['Weight'].mean().round(1)} kg.")
print(f"The mean weight for those without Super Speed is {super_speed_false['Weight'].mean().round(1)} kg.")

The mean weight for those with Super Speed is 129.4 kg.
The mean weight for those without Super Speed is 101.8 kg.


## Question 2

In [44]:
print(f"The mean height (in cm) of heroes grouped by publisher are:\n\n{super_hero_df.groupby('Publisher')['Height'].mean().round(1)}")

The mean height (in cm) of heroes grouped by publisher are:

Publisher
DC Comics            181.9
Dark Horse Comics    176.9
George Lucas         159.6
Image Comics         211.0
Marvel Comics        191.5
Shueisha             171.5
Star Trek            181.5
Team Epic TV         180.8
Unknown              178.0
Name: Height, dtype: float64
